lib/actions/appleMusic.js

```diff
@@ -155,15 +145,16 @@ function appleMusic(player, values) {
                 return player.coordinator.clearQueue()
                   .then(() => player.coordinator.addURIToQueue(queueTracks[0].uri, queueTracks[0].metadata, true, 1))
                   .then(() => {
-                      if (radioOn(player)) {
-                        return promise.then(() => player.coordinator.setAVTransport(queueURI,''));
+                      if (isRadioOrLineIn(player.coordinator.state.currentTrack.uri)) {
+                        player.coordinator.setAVTransport(queueURI,''));
                       }
+                      return Promise.resolve();
                     })
                   .then(() => player.coordinator.play())
                   .then(() => {
                     return queueTracks.reduce(function(promise, track, index){
                       if (index > 0) {
-                        return promise.then(() => player.coordinator.addURIToQueue(track.uri, track.metadata, true, index + 1));
+                        player.coordinator.addURIToQueue(track.uri, track.metadata, true, index + 1);
                       } else {
                         return promise.then(() => {});
                       }   
```

Comment:
Very helpful.  I moved to the dark side (management) full time after a lot of C/C++.  Played around with a little Java and JavaScript (now), but never C#.  I recognized the => , but I couldn’t remember where I saw this and it was throwing me a little.  The first part of the documentation answered the question and is what I was starting to figure out.  Thanks for the link.

I had just gotten use to callback hell and now I’m still trying to get my head around Promises.  Getting close, but not quite there yet.

Thanks again…


[cid:image001.gif@01D1C246.AEA36C60]<http://www.infor.com/>

Jim Plourde | VP Cloud Services
mobile: 864.270.7677 | office: 864.546.5089 | jim.plourde@infor.com<mailto:jim.plourde@infor.com> | http://www.infor.com<http://www.infor.com/>



From: Jimmy Shimizu <notifications@github.com>
Reply-To: jishi/node-sonos-http-api <reply@reply.github.com>
Date: Thursday, June 9, 2016 at 11:41 AM
To: jishi/node-sonos-http-api <node-sonos-http-api@noreply.github.com>
Cc: James Plourde <Jim.Plourde@infor.com>, Author <author@noreply.github.com>
Subject: Re: [jishi/node-sonos-http-api] Update is-radio-or-line-in.js (#235)


In lib/actions/appleMusic.js<https://github.com/jishi/node-sonos-http-api/pull/235#discussion_r66465881>:

>                      })

>                    .then(() => player.coordinator.play())

>                    .then(() => {

>                      return queueTracks.reduce(function(promise, track, index){

>                        if (index > 0) {

> -                        return promise.then(() => player.coordinator.addURIToQueue(track.uri, track.metadata, true, index + 1));

> +                        player.coordinator.addURIToQueue(track.uri, track.metadata, true, index + 1);

>                        } else {

>                          return promise.then(() => {});

>                        }

I get the confusion.

1.      This is not actually related to Promises, but rather arrow functions which are similar to arrow functions in C# (and perhaps Java as well). https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions. Arrow functions without body (basically single-line without the {}) are direct returning.

2.      First argument to the reduce delegate/callback is always the value returned from the previous run. You don't have an initial value (which is the second argument to reduce(), which means that it will be the first value in the array instead (which isn't a promise). So this won't work:

[1,2,3].reduce((promise, value) => promise.then())

but this will:

[1,2,3].reduce((promise, value) => promise.then(), Promise.resolve())

In the first example, promise will be the literal value 1 at first run, and value will be 2.
In the second example, it will be a simple promise, because of the second argument to reduce. value will 1 at first run.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/jishi/node-sonos-http-api/pull/235/files/e541c8788ccd05ff5f213aed7027e777da5a6b38#r66465881>, or mute the thread<https://github.com/notifications/unsubscribe/ASedkKvYEC-3ourlhTtPMX_389iIUVQqks5qKDQagaJpZM4Ixe4I>.